[AMDGPU] Promote uniform ops to I32 in DAGISel

Pierre-vh opened the pull request:
Promote uniform binops, selects and setcc between 2 and 16 bits to 32 bits in DAGISel

Solves #64591

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUCombine.td:
```diff
@@ -145,6 +145,31 @@ def expand_promoted_fmed3 : GICombineRule<
 
 } // End Predicates = [NotHasMed3_16]
 
+def promote_i16_uniform_binops_frag : GICombinePatFrag<
```
I'll look at moving this into the legalizer

Pierre-vh replied:
I just remembered that the legalizer doesn't have uniformity info. This feels more like a legalization rule but it looks like it'll be stuck here due to the need for uniformity info, unless I'm missing another way to implement it?

arsenm replied:
It's not really legalization. It's a register bank aware optimization. I would drop all the globalisel handling from this patch, it's a separate issue

Pierre-vh replied:
This will switch the CGP i16 promotion off, so we'd let GISel regress in the meantime?

arsenm replied:
I'd first implement all the DAG parts in one patch and leave the IR pass as it is. Then the next 2 parts can be done in either order, either do the globalisel parts or remove the IR promotion and let the regression happen

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6746,6 +6748,122 @@ SDValue SITargetLowering::lowerFLDEXP(SDValue Op, SelectionDAG &DAG) const {
   return DAG.getNode(ISD::FLDEXP, DL, VT, Op.getOperand(0), TruncExp);
 }
 
+static unsigned getExtOpcodeForPromotedOp(SDValue Op) {
```
I would hope there would be a generic helper for this somewhere

Pierre-vh replied:
I haven't found one, there are some `getExtendFor` functions but they're not for binops

Pierre-vh committed:
[AMDGPU] Promote uniform ops to I32 in ISel

Promote uniform binops, selects and setcc in Global & DAGISel instead of CGP.

Solves #64591

jayfoad reviewed (commented):
Some of my comments on #106557 also apply here.

Pierre-vh committed:
Address coments

Pierre-vh committed:
rebase

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -1022,14 +1022,45 @@ bool AMDGPUTargetLowering::isZExtFree(EVT Src, EVT Dest) const {
+  case ISD::SETCC:
+  case ISD::SELECT:
+    if (Subtarget->has16BitInsts() &&
+        (DestVT.isVector() ? !Subtarget->hasVOP3PInsts() : true)) {
```
DestVT.isVector() && !Subtarget->hasVOP3PInsts()

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6733,6 +6735,107 @@ SDValue SITargetLowering::lowerFLDEXP(SDValue Op, SelectionDAG &DAG) const {
+  if (DCI.isBeforeLegalizeOps())
+    return SDValue();
+
+  // Promote only if:
+  //    - We have 16 bit insts (not true 16 bit insts).
+  //    - We don't have packed instructions (for vector types only).
+  // TODO: For vector types, the set of packed operations is more limited, so
+  // may want to promote some anyway.
+  if (!Subtarget->has16BitInsts() ||
+      (OpTy.isVector() ? Subtarget->hasVOP3PInsts() : false))
+    return SDValue();
+
+  // Promote uniform scalar and vector integers between 2 and 16 bits.
+  if (Op->isDivergent() || !OpTy.isInteger() ||
+      OpTy.getScalarSizeInBits() == 1 || OpTy.getScalarSizeInBits() > 16)
+    return SDValue();
+
```
This is repeating isNarrowingProfitable, can you just call it?

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6733,6 +6735,107 @@ SDValue SITargetLowering::lowerFLDEXP(SDValue Op, SelectionDAG &DAG) const {
+  // truncate back to the original type.
+  SDValue NewVal;
+  if (Opc == ISD::SELECT)
+    NewVal = DAG.getSelect(DL, ExtTy, Op->getOperand(0), LHS, RHS);
```
Should use getNode(ISD::SELECT, getSelect may swap you to a VSELECT depending on the type

Pierre-vh committed:
Comments

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgpu-simplify-libcall-pow-codegen.ll:
```diff
@@ -18,189 +18,33 @@ declare hidden half @_Z4pownDhi(half, i32)
 
 define half @test_pow_fast_f16(half %x, half %y) {
-; CHECK-LABEL: test_pow_fast_f16:
-; CHECK:       ; %bb.0:
```
This file lost all the test checks

arsenm replied:
I fixed this in 528bcf3a55ca520c31c77ed5fbacf09bff8f39ec

jayfoad commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6733,6 +6735,93 @@ SDValue SITargetLowering::lowerFLDEXP(SDValue Op, SelectionDAG &DAG) const {
+  LHS = DAG.getNode(ExtOp, DL, ExtTy, {LHS});
+
+  // Special case: for shifts, the RHS always needs a zext.
+  if (Op.getOpcode() == ISD::SRA || Op.getOpcode() == ISD::SRL ||
+      Op.getOpcode() == ISD::SRA)
```
```suggestion
  if (Op.getOpcode() == ISD::SHL || Op.getOpcode() == ISD::SRL ||
      Op.getOpcode() == ISD::SRA)
```

jayfoad replied:
@Pierre-vh ping - this looks like it was a simple typo?

Pierre-vh replied:
Oops sorry, I'll fix it right now
